pkg/admin/migration/migration.go

```diff
@@ -0,0 +1,282 @@
+			continue
+		}
+		if customGroups.Has(gv.Group) {
+			klog.V(4).Infof("ignored group %v because it's a custom group", gv.Group)
+			continue
+		}
+		if aggregatedGroups.Has(gv.Group) {
+			klog.V(4).Infof("ignored group %v because it's an aggregated group", gv.Group)
+			continue
+		}
+		for _, r := range resourceList.APIResources {
+			// ignore subresources
+			if strings.Contains(r.Name, "/") {
+				continue
+			}
+			if blackListResources.Has(r.Name) {
+				continue
+			}
+			// ignore resources that cannot be listed and updated
+			if !sets.NewString(r.Verbs...).HasAll("list", "update") {
+				continue
+			}
+			gvs := resourceToGroupVersions[r.Name]
+			gvs = append(gvs, gv)
+			resourceToGroupVersions[r.Name] = gvs
+		}
+	}
+
+	ret := []schema.GroupVersionResource{}
+	for resource, groupVersions := range resourceToGroupVersions {
+		if len(groupVersions) == 1 {
+			continue
+		}
+		ret = append(ret, groupVersions[0].WithResource(resource))
+	}
+	return ret, nil
+}
+
+func (m *migrator) migrateOneItem(ctx context.Context, resource schema.GroupVersionResource, item *unstructured.Unstructured) error {
+	namespace, err := metadataAccessor.Namespace(item)
+	if err != nil {
+		return err
+	}
+	name, err := metadataAccessor.Name(item)
+	if err != nil {
+		return err
+	}
+	getBeforePut := false
+	for {
+		getBeforePut, err = m.try(ctx, resource, namespace, name, item, getBeforePut)
+		if err == nil || errors.IsNotFound(err) {
+			return nil
+		}
+		if canRetry(err) {
+			seconds, delay := errors.SuggestsClientDelay(err)
+			switch {
+			case delay && len(namespace) > 0:
+				klog.Warningf("migration of %s, in the %s namespace, will be retried after a %ds delay: %v", name, namespace, seconds, err)
+				time.Sleep(time.Duration(seconds) * time.Second)
+			case delay:
+				klog.Warningf("migration of %s will be retried after a %ds delay: %v", name, seconds, err)
+				time.Sleep(time.Duration(seconds) * time.Second)
+			case !delay && len(namespace) > 0:
+				klog.Warningf("migration of %s, in the %s namespace, will be retried: %v", name, namespace, err)
+			default:
+				klog.Warningf("migration of %s will be retried: %v", name, err)
+			}
+			continue
+		}
+		// error is not retriable
+		return fmt.Errorf("can not retry: %+v", err)
+	}
+}
+
+func (m *migrator) try(ctx context.Context, resource schema.GroupVersionResource, namespace, name string, item *unstructured.Unstructured, get bool) (bool, error) {
+	var err error
+	if get {
+		item, err = m.client.
+			Resource(resource).
+			Namespace(namespace).
+			Get(ctx, name, metav1.GetOptions{})
+		if err != nil {
+			return true, err
+		}
+	}
+	_, err = m.client.
+		Resource(resource).
+		Namespace(namespace).
+		Update(ctx, item, metav1.UpdateOptions{})
+	if err == nil {
+		return false, nil
+	}
+	return errors.IsConflict(err), err
+}
+
+func (m *migrator) list(ctx context.Context, resource schema.GroupVersionResource, options metav1.ListOptions) (*unstructured.UnstructuredList, error) {
```

Comment:
This function appears to be used only once, do we need it?

Reply:
Nah, we can get rid of it